src/connection-manager/topology.js

```diff
@@ -0,0 +1,102 @@
+'use strict'
+
+class Topology {
+  /**
+   * @param {Object} props
+   * @param {number} props.min minimum needed connections (default: 0)
+   * @param {number} props.max maximum needed connections (default: Infinity)
+   * @param {function} props.onConnect protocol "onConnect" handler
+   * @param {function} props.onDisconnect protocol "onDisconnect" handler
+   * @param {Array<string>} props.multicodecs protocol multicodecs
+   * @param {Registrar} registrar
+   * @constructor
+   */
+  constructor ({
+    min = 0,
+    max = Infinity,
+    onConnect,
+    onDisconnect,
+    multicodecs,
+    registrar
+  }) {
+    this.multicodecs = multicodecs
+    this.registrar = registrar
+    this.min = min
+    this.max = max
+    this.peers = new Map()
+
+    // Handlers
+    this._onConnect = onConnect
+    this._onDisconnect = onDisconnect
+
+    this._onProtocolChange = this._onProtocolChange.bind(this)
+
+    // Set by the registrar
+    this._peerStore = null
+  }
+
+  /**
+   * Set peerstore to the topology.
+   * @param {PeerStore} peerStore
+   */
+  set peerStore (peerStore) {
+    this._peerStore = peerStore
+
+    this._peerStore.on('change:protocols', this._onProtocolChange)
+  }
+
+  /**
+   * Try to add a connected peer to the topology, if inside the thresholds.
+   * @param {PeerInfo} peerInfo
+   * @param {Connection} connection
+   * @returns {void}
+   */
+  tryToConnect (peerInfo, connection) {
+    // TODO: conn manager should validate if it should try to connect
+
+    this._onConnect(peerInfo, connection)
+
+    this.peers.set(peerInfo.id.toB58String(), peerInfo)
+  }
+
+  /**
+   * Notify protocol of peer disconnected.
+   * @param {PeerInfo} peerInfo
+   * @param {Error} [error]
+   * @returns {void}
+   */
+  disconnect (peerInfo, error) {
+    if (this.peers.delete(peerInfo.id.toB58String())) {
+      this._onDisconnect(peerInfo, error)
+    }
+  }
+
+  /**
+   * Check if a new peer support the multicodecs for this topology.
+   * @param {Object} props
+   * @param {PeerInfo} props.peerInfo
+   * @param {Array<string>} props.protocols
+   */
+  _onProtocolChange ({ peerInfo, protocols }) {
+    const existingPeer = this.peers.get(peerInfo.id.toB58String())
+
+    protocols.filter(protocol => this.multicodecs.includes(protocol))
+
+    // Not supporting the protocol anymore?
+    if (existingPeer && protocols.filter(protocol => this.multicodecs.includes(protocol)).length === 0) {
+      this._onDisconnect({
+        peerInfo
+      })
+    }
+
+    // New to protocol support
+    for (const protocol of protocols) {
+      if (this.multicodecs.includes(protocol)) {
+        this.tryToConnect(peerInfo, this.registrar.getPeerConnection(peerInfo))
```

Comment:
What's the point of `tryToConnect`? Ultimately I think this line should just be an add to `this.peers`. When registration happens we could use `Proxy` to have the registrar watch for changes on each Toplogies peer set. When that changes it can take action on prioritizing peers for the connection manager to connect/disconnect to.

The flow I am thinking of is:
1. A new peer is discovered and we are under the min threshold for a topology (multicodec, bootstrap, root, etc)
1. The connection manager takes the highest priority peers and dials to them (for a typical node on startup, this would be the bootstrap peers, and any other priority peers like ipfs preload nodes)
1. When the connection is made, the registrar is notified and in turn notifies the topologies, so they can take action if desired
  a. A multicodec topology could just ignore this, because it cares about protocol updates
1. When identify runs, the protocol change is triggered by the peer store, and the multicodec does a check for its multicodecs
  a. If the multicodec is present, it adds the peer to its list of peers
  b. The peer set change, causes the registrar to update the priority of that peer
1. The reprioritized peer will now be connected/disconnected accordingly by the connection manager

Reply:
The point for `tryToConnect` is basically the following use case:

When a peer starts a subsystem (`pubsub` as an example), the peer may already be connected to several peers. Take into account that someone enables the subsystem in runtime. By that time, the peer will know and be connected to several peers (some of them may support the protocol of the subsystem). Once the subsystem is started and the registrar is notified of it, the registrar must inform the topology that it is already connected with some peers (`tryToConnect`), and the topology will do `onConnect` if it is inside the thresholds. This way, the `tryToConnect` is basically a `subsystem.onConnect` with a validation before, in order to guarantee that we are inside the thresholds.

The flow that you described looks good, except for the initial part that I just described. However, I am not sure if we are getting too complex with the `Proxy`.

Reply:
>  the registrar must inform the topology that it is already connected with some peers

I would expect that when a topology is registered, the registrar would simply trigger `onConnect` for any existing connections. It would basically be a replay mechanism for lazily started systems.

> However, I am not sure if we are getting too complex with the Proxy

It's not really complex though, it's just another way of creating an observer. We could do it with hooks or events, but the Proxy is cleaner IMO. The registrar should be the only thing observing `topology.peers`.

Reply:
> I would expect that when a topology is registered, the registrar would simply trigger onConnect for any existing connections. It would basically be a replay mechanism for lazily started systems.

I see 2 problems with that. The topology needs to update its `peers` set and if we are already connected with a lot of peers, we may immediately exceed the threshold. That's why I would rather have a `tryToConnect` method.

> It's not really complex though, it's just another way of creating an observer. We could do it with hooks or events, but the Proxy is cleaner IMO. The registrar should be the only thing observing topology.peers.

I think we can iterate on that later, as we will not need this until getting to connectionManager